src/Pool.sol

```diff
@@ -141,157 +157,243 @@ abstract contract Pool is Pausable, Ownable, ERC4626 {
+            uint128 depositAmount_ = _usersData[msg.sender].deposits[i].amount;
+
+            // Compute the withdraw fee and the resulting amount of the deposit.
+            uint64 withdrawFee_ =
+                uint64(_calcWithdrawFeePerBlock(_usersData[msg.sender].deposits[i].blockNumber) * MULTIPLIER / 100);
+            uint128 depositAmountFeed_ = uint128(depositAmount_ - depositAmount_ * withdrawFee_ / MULTIPLIER);
+
+            // Check if the deposit can manage the full remaining amount of the withdrawal.
+            if (depositAmountFeed_ > amountRem_) {
+                // Update user information regarding the deposit and the total amount deposited.
+                _usersData[msg.sender].deposits[i].amount -= amountRem_;
+                _usersData[msg.sender].totalAmountDeposited -= amountRem_;
+
+                // Update the withdrawAmount and the totalFee.
+                withdrawAmount_ += amountRem_;
+                totalFee_ += uint128((amountRem_ * withdrawFee_) / MULTIPLIER);
+
+                continue;
+            }
+
+            // Update user information regarding the deposit and the total amount deposited.
+            _usersData[msg.sender].deposits[i].amount = 0;
+            _usersData[msg.sender].totalAmountDeposited -= depositAmount_;
+
+            // Update the total deposited amount.
+            _totalAmount -= depositAmount_;
+
+            // Update the withdrawAmount and the totalFee.
+            withdrawAmount_ += depositAmountFeed_;
+            totalFee_ += uint128((depositAmount_ * withdrawFee_) / MULTIPLIER);
+
+            // This deposit it's no longer available. Update the nextDeposit to be removed.
+            _usersData[msg.sender].nextDepositIdToRemove = i == _usersData[msg.sender].deposits.length - 1 ? i : i + 1;
+        }
+
+        _depositAsset.approve(address(this), withdrawAmount_ + totalFee_);
+        // Transfer the assets to the user.
+        bool success_ = _depositAsset.transferFrom(address(this), msg.sender, withdrawAmount_);
+        if (!success_) revert Pool_AssetTransferFailedError();
+
+        // Transfer the withdraw fee into the recipient address.
+        success_ = _depositAsset.transferFrom(address(this), _depositFeeRecipient, totalFee_);
+        if (!success_) revert Pool_WithdrawFeeChargeFailedError();
     }
 
     /**
-     * @dev Claims rewards for a specific deposit.
-     * @param depositId ID of the deposit.
+     * @dev Claim rewards for a specific deposit.
      */
-    function claimRewards(uint32 depositId) external {
-        /*
-            1. Checks if the deposit exists.
-            2. Check if there is any active Booster Pack.
-            3. Set accumRewards to 0.
-            4. Transfer the rewards to the depositor.
-        */
+    function claimRewards() external nonReentrant {
+        // Update the rewards per token.
+        updateRewardsPerToken();
+
+        // Check user's deposits in order to compute the total rewards accumulated.
```

Comment:
critical: the accumulated rewards must be per deposit such that when it is removed, its accumulated rewards are also removed.
example: 5 rewardsPerToken * 10 amountDeposited = 50 -> withdraw 5 amount deposited.
then 6 rewardsPerToken * 5 amountDeposited = 30.
The next calculus will be 30 - 50 and will revert.